x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,140 @@
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isIPAndExact;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isInteger;
+
+/**
+ * Converts an IP value and a prefix length to the prefix.
+ */
+public class IpPrefix extends EsqlScalarFunction {
+
+    private final Expression ipField;
+    private final Expression prefixLengthField;
+
+    @FunctionInfo(
+        returnType = "ip",
+        description = "Returns the original IP with all but the first N bits set to zero.",
+        examples = @Example(file = "ip", tag = "ipPrefix")
+    )
+    public IpPrefix(
+        Source source,
+        @Param(
+            name = "ip",
+            type = { "ip" },
+            description = "IP address of type `ip` (both IPv4 and IPv6 are supported)."
+        ) Expression ipField,
+        @Param(
+            name = "prefixLength",
+            type = { "integer" },
+            description = "Prefix length. If in the range (0, 32), the IP is treated as an IPv4 address. "
+                + "If in the range (32, 128), the IP is treated as an IPv6 address. "
+                + "If the prefix length is out of range, the function returns `null`."
+        ) Expression prefixLengthField
+    ) {
+        super(source, Arrays.asList(ipField, prefixLengthField));
+        this.ipField = ipField;
+        this.prefixLengthField = prefixLengthField;
+    }
+
+    public Expression ipField() {
+        return ipField;
+    }
+
+    public Expression prefixLengthField() {
+        return prefixLengthField;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var ipEvaluatorSupplier = toEvaluator.apply(ipField);
+        var prefixLengthEvaluatorSupplier = toEvaluator.apply(prefixLengthField);
+        return new IpPrefixEvaluator.Factory(source(), ipEvaluatorSupplier, prefixLengthEvaluatorSupplier);
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef ip, int prefixLength) {
+        if (ip.length != 16 || prefixLength < 0 || prefixLength > ip.length * 8) {
+            return null;
+        }
+
+        boolean isIpv4 = prefixLength <= 32;
+        // IPv4 is stored in the last 4 bytes
+        int offset = isIpv4 ? 12 : 0;
+        int fullBytes = offset + prefixLength / 8;
+        int remainingBits = prefixLength % 8;
+
+        BytesRef prefix = new BytesRef(new byte[ip.length]);
```

Comment:
If you pass a `ByteRef` in as a `@Fixed` then you don't have to allocate it on every iteration. It's a silly hack to make java not do so many allocations, but it'll get the job done here I think.

Reply:
And it's safe to return the same `BytesRef` from the function over and over again - the contract for these is that you always make a copy if you need to keep the bytes. Usually you don't though.

Reply:
Added the Fixed BytesRef. Please check if that's how it's expected to be use.
In all the cases I found of ByteRefs, an empty ByteRef was supplied. I'm supplying instead a BytesRef with an internal bytes array with the required size, as otherwise, being BytesRef a lightweight wrapper, it didn't make sense to me to just reuse it (?)

About the contract of functions, from what you tell me, I understand that the return won't be mutated outside of your function (And callers will always clone it if they need to keep the return), and that you shouldn't mutate the parameters, right?